raiden/raiden_service.py

```diff
@@ -465,19 +520,39 @@ def on_message(self, message, msghash):
             self.message_confirmtransfer(message)
 
         else:
-            raise Exception("Unknow cmdid '{}'.".format(cmdid))
+            raise Exception("Unhandled message cmdid '{}'.".format(cmdid))
 
-    def message_ping(self, message):
+    def message_ping(self, message):  # pylint: disable=unused-argument,no-self-use
         log.info('ping received')
 
     def message_confirmtransfer(self, message):
         pass
 
+    def message_revealsecret(self, message):
+        self.raiden.message_for_task(message, message.hashlock)
+
     def message_secretrequest(self, message):
         self.raiden.message_for_task(message, message.hashlock)
 
     def message_secret(self, message):
-        self.raiden.message_for_task(message, message.hashlock)
+        # notify the waiting tasks about the message (multiple tasks could
+        # happen in exchanges were a node is on both asset transfers), and make
+        # sure that the secret is register if it fails (or it has exited
+        # because the transfer was considered done)
+        try:
+            self.raiden.message_for_task(message, message.hashlock)
+        finally:
+            try:
+                # register the secret with all channels interested in it (this
+                # must not withdraw or unlock otherwise the state changes could
+                # flow in the wrong order in the path)
+                self.raiden.register_secret(message.secret)
+
+                # make sure we have a proper state change
+                asset_manager = self.raiden.get_manager_by_asset_address(message.asset)
+                asset_manager.handle_secretmessage(message)
+            except:  # pylint: disable=bare-except
+                log.exception('Unhandled exception')
```

Comment:
I believe we should at least log the content/type of the unhandled exception here.